perf: use hashing to determine the format in infer/lib/audio.py

413x1nkp opened the pull request:
this makes use of python's dictionaries, comprehension of which is much more performant than the usage of if-elif-else 
- added type-safety for the functions
- added additional check on whether or not the file exists to not attempt to create a pipe if it doesn't (mitigates expensive subprocess call)

413x1nkp committed:
perf: use hashing to determine the format

this makes use of python's dictionaries, comprehension of which is much more performant than the usage of if-elif-else blocks

413x1nkp committed:
chore: use pathlib for crossplatform path transformations in audio.py

413x1nkp committed:
chore: add type safety and implement file.exists() check for the load_audio function

413x1nkp committed:
Merge branch 'dev' into optim-audio

413x1nkp committed:
fix: use typing's `Dict` instead of Python's `dict` for Python<=3.9

413x1nkp committed:
fix: don't use type subscription for compatibility with Python<3.9

413x1nkp commented:
Please squash in case of PR being accepted! I'm using Python3.12 as a primary python version, some things like type subscription simply don't work in some of the tests intended for Python<3.9, hence the multiple `fix` commits

413x1nkp committed:
Merge branch 'dev' into optim-audio

fumiama commented on infer/lib/audio.py:
```diff
@@ -1,18 +1,24 @@
     inp = av.open(i, "r")
-    if format == "m4a":
-        format = "mp4"
+    format = video_format_dict[format]
```
I think the dict will panic if the format is not m4a. You can use the `get` grammar and set the default value as `format` itself.

413x1nkp replied:
Should be fixed by `f145e7c`

fumiama commented on infer/lib/audio.py:
```diff
@@ -1,18 +1,24 @@
-        format = "libvorbis"
-    if format == "mp4":
-        format = "aac"
+    format = audio_format_dict[format]
```
Same issue.

413x1nkp replied:
Should be fixed by `f145e7c`

413x1nkp committed:
fix: use `.get()` for format dictionaries, also provide default values in case `.get()` yields `None`

fumiama commented on infer/lib/audio.py:
```diff
@@ -16,9 +16,9 @@
 def wav2(i: BytesIO, o: BufferedWriter, format: str):
     inp = av.open(i, "r")
-    format = video_format_dict[format]
+    format = video_format_dict.get(format) or "mp4"
```
Well, just `get(format, format)`, if my memory is correct.

413x1nkp replied:
Huh, I didn't know Python supported that
Will change.

413x1nkp committed:
fix: set default format via argument to `.get()` method

this is used in favor of providing the default as `.get(format) or <DEFAULT>`

413x1nkp committed:
Merge branch 'dev' into optim-audio

fumiama commented on infer/lib/audio.py:
```diff
@@ -1,18 +1,24 @@
     inp = av.open(i, "r")
-    if format == "m4a":
-        format = "mp4"
+    format = video_format_dict.get(format, "mp4")
```
Well, well, just `get(format, format)` 😂. You can run a test on this grammar.

413x1nkp replied:
oh... OHHHHHH
Sorry :sweat_smile:

413x1nkp committed:
fix: default to the user-requested format in case of hashing failure

413x1nkp committed:
fix: don't import `Any` from typing module since a type subscription was reverted previously

`Any` was used here for type subscription, but since now it's just `np.ndarray` and not `np.ndarray[Any, ...]` there's no need to import Any.

fumiama reviewed (approved):
You can see that this file used two ffmepg impls, one is av, the other is ffmpeg. I would like to replace all the ffmpeg usage (not so much, just one or two) to av so that We can drop the need to download ffmpeg at the root dir or path. If you are also interesting about this, you can have a try.

413x1nkp commented:
> You can see that this file used two ffmepg impls, one is av, the other is ffmpeg. I would like to replace all the ffmpeg usage (not so much, just one or two) to av so that We can drop the need to download ffmpeg at the root dir or path. If you are also interesting about this, you can have a try.

That's a great idea, to be honest! I'll try to do that, but a lot of the audio-related operations are all so... decentralized right now
But I'm willing to give it a try! :heart:

fumiama commented:
> > You can see that this file used two ffmepg impls, one is av, the other is ffmpeg. I would like to replace all the ffmpeg usage (not so much, just one or two) to av so that We can drop the need to download ffmpeg at the root dir or path. If you are also interesting about this, you can have a try.
> 
> 
> 
> That's a great idea, to be honest! I'll try to do that, but a lot of the audio-related operations are all so... decentralized right now
> 
> But I'm willing to give it a try! :heart: 

You can use the search function in your IDE and just search the string `ffmpeg.` (donnot forget the last dot). The result is all the usage. One is just in this file, the other is a call to `ffprobe`.

413x1nkp commented:
> You can use the search function in your IDE and just search the string `ffmpeg.` (donnot forget the last dot). The result is all the usage. One is just in this file, the other is a call to `ffprobe`.

Created a PR introducing this change. #31
